### PR TITLE
Migrate Lisk Modular repository files to framework directory and create lisk-chain module  - Closes #2763, 2762

### LINKED ISSUE #2762: Restructure the file directories for components, modules and framework files
### Expected behavior
The top level directory structure should be 
```
/framework
    src/
        /components
            /storage
            /cache
            /logger
            /system
        /modules
            /chain
            /http_api
        controller.js
        bus.js
        application.js
/src
    /config
    index.js
/docker
/build
/test
    framework/
package.json
```
Same structure should be followed in the `test` directory as well. Later when decided we will make each sub directory under `components`, `modules` and `framework` as seperate `npm package` choose to convert it to a mono repo or move them to seperate repositories. 

### Actual behavior
Currently files directories are distributed among `modules`, `logic`, `helpers` directories, this structure is dedicated for `chain` module. 

### Which version(s) does this affect? (Environment, OS, etc...)
`1.6`

### LINKED ISSUE #2763: Migrate Lisk Modular repository files to framework directory and create lisk-chain module
Need to move all framework related files to `lisk-modular` repository the `lisk` repository under `/framework` directory. 

After, move all code of lisk into a module `lisk-chain` and load it with the controller with minimum changes possible (https://github.com/LiskHQ/lisk/issues/2761).

### PR DESCRIPTION
**Pull request by nazarhussain**
### What was the problem?
File structure was aligned to old relates. 

### How did I fix it?

For the newer release `1.6` need to update the file structure and implemented the lisk-modular concept `with minimal changes` possible.

### Review checklist

* The PR resolves #2762, #2763
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: framework/src/controller/bus.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+				throw new Error(`Action "${actionName}" already registered with bus.`);
+			}
+			this.actions[actionName] = true;
+		});
+	}
+
+	invoke(actionData) {
+		const action = Action.deserialize(actionData);
+
+		if (action.module === 'lisk') {
```
Comment (MaciejBaj) [inline framework/src/controller/bus.js]: Why is 'lisk' value hardcoded? Should be passed by configuration as app name.

File: framework/src/controller/event.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	 * @param data - Data associated with the event
+	 */
+	constructor(name, data = null, source = null) {
+		assert(
+			eventWithModuleNameReg.test(name),
+			`Event name "${name}" must be a valid name with module name.`
+		);
+
+		const matches = eventWithModuleNameReg.exec(name);
+		// eslint-disable-next-line prefer-destructuring
```
Comment (MaciejBaj) [inline framework/src/controller/event.js]: You can actually use destructuring in the line below.

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+				...config.redis,
+				enabled: config.cacheEnabled,
+			},
+			storage: config.db,
+		},
+	});
+
+	app.overrideModuleConfig('chain', { exceptions, config });
+
+	app.run();
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply src/index.js]: Yes that' possible, but we still need to wrap the whole code for any sync error in application constructor.

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+const _ = require('lodash');
+const { Application } = require('../framework/src');
+const genesisBlock = require('../config/devnet/genesis_block');
+const constants = require('../config/devnet/constants');
+const exceptions = require('../config/devnet/exceptions');
+const defaultConfig = require('../config/default/config');
+const networkConfig = require('../config/devnet/config');
+
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply src/index.js]: I am maintaining all issues locally, now add all those to github.

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	(objValue, srcValue) => {
+		if (_.isArray(objValue)) {
+			return srcValue;
+		}
+		return undefined;
+	}
+);
+
+try {
+	const app = new Application('devnet', genesisBlock, constants, {
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply src/index.js]: Constants I believe are something shared/injected among all modules. Default constants will be kept by application and here in constructor you can pass overriding constants only.

File: framework/src/controller/application.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+			`A transaction with alias "${alias}" already registered.`
+		);
+
+		const transactions = this.getTransactions();
+		transactions[alias] = Object.freeze(Transaction);
+		scope.transactions.set(this, transactions);
+	}
+
+	getTransactions() {
+		return scope.transactions.get(this);
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply framework/src/controller/application.js]: To keep the private data members of an instance. We already have a lot of discussion on this topic and that's a preferable way in JS/ES6 by the community.

File: framework/src/controller/action.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+const actionWithModuleNameReg = /^[a-zA-Z][a-zA-Z0-9]*:[a-zA-Z][a-zA-Z0-9]*$/;
+
+module.exports = class Action {
+	/**
+	 *
+	 * @param eventName - Can be simple event or be combination of module:event
+	 * @param {array} params - Params associated with the action
+	 * @param moduleName - Module name if event name does not have its prefix
+	 */
+	constructor(name, params = null, source = null) {
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply framework/src/controller/action.js]: Will updated all JSDocs

File: framework/src/controller/bus.js (hunk 1 above)
Comment (nazarhussain) [reply framework/src/controller/bus.js]: That's an identifier for controller, yes it most be moved to a constant and will do it. First we need to define what is app name is defining, if that identifier remains same among networks of same app then we can use it. If not then we have to stick with one.

**Comment by nazarhussain** [comment]
@MaciejBaj Related to your recent commits, I don't think putting transactions files on root directory of repo is a good idea. 

Only dapp, in_transfer, out_transfer belongs to lisk-core implementation. All others are part of the actual framework and chain module.

File: framework/src/controller/controller.js [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+		this.channel.publish('lisk:ready', {});
+	}
+
+	// eslint-disable-next-line class-methods-use-this
+	async _setupDirectories() {
+		// Make sure all directories exists
+		fs.emptyDirSync(config.dirs.temp);
+		fs.ensureDirSync(config.dirs.sockets);
+		fs.ensureDirSync(config.dirs.pids);
+		fs.writeFileSync(`${config.dirs.pids}/controller.pid`, process.pid);
```
[the remark this answers is not in the document]
Comment (yatki) [reply framework/src/controller/controller.js]: We also must keep this in mind when implementing RPC communication. Sockets folder name itself or the socket names must be unique per instance.

[the remark this answers is not in the document]
Comment (nazarhussain) [reply framework/src/controller/controller.js]: Yes it is, there is some discussions meetings already initiated on this topic.

**Comment by nazarhussain** [comment]
@diego-G All tests are green now.

File: framework/src/controller/application.js [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+const defaults = require('./defaults');
+const version = require('../version');
+const validator = require('./helpers/validator');
+const schema = require('./schema/application');
+const { createLoggerComponent } = require('../components/logger');
+
+const ChainModule = require('../modules/chain');
+
+// Private scope used because private keyword is restricted
+const scope = {
```
Comment (shuse2) [inline framework/src/controller/application.js]:
```suggestion
const private = {
```

File: framework/src/controller/channels/base.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	'loading:finished',
+];
+
+/**
+ * Base Channel class responsible which used as reference to implement others channels for bus to module communication
+ *
+ * @namespace Framework.channels
+ * @type {module.Base}
+ */
+module.exports = class Base {
```
Comment (shuse2) [inline framework/src/controller/channels/base.js]:
```suggestion
module.exports = class BaseChannel {
```

File: framework/src/controller/application.js (hunk 9 above)
Comment (nazarhussain) [reply framework/src/controller/application.js]: I tried private before, but eslint was complaining.

Comment (shuse2) [reply framework/src/controller/application.js]: do we need to put this in nested object?

File: framework/src/controller/application.js [hunk 11]
```diff
@@ -8,8 +8,8 @@ const { createLoggerComponent } = require('../components/logger');
 
 const ChainModule = require('../modules/chain');
 
-// Private scope used because private keyword is restricted
-const scope = {
+// Private __private used because private keyword is restricted
+const __private = {
```
Comment (shuse2) [inline framework/src/controller/application.js]: storage is using `_` convention, so maybe just single underscore is ok?

File: framework/src/controller/application.js (hunk 9 above)
Comment (yatki) [reply framework/src/controller/application.js]: @shuse2 having a nested object increases code readability I think. It helps to differentiate the private variables from global/public variables.

File: framework/src/controller/application.js (hunk 11 above)
Comment (nazarhussain) [reply framework/src/controller/application.js]: @shuse2 We are using `__` syntax already in the code everywhere. 

https://github.com/LiskHQ/lisk/blob/cb68ebb96aacc3dccd532de6f926a1c39a2b1411/modules/loader.js#L30
https://github.com/LiskHQ/lisk/blob/cb68ebb96aacc3dccd532de6f926a1c39a2b1411/modules/blocks/verify.js#L35

The convention of `_` was used for parameters which are not used in the functions.

File: framework/src/components/logger/logger.js [hunk 12]
```diff
@@ -0,0 +1,151 @@
+/*
```
[the remark this answers is not in the document]
Comment (diego-G) [reply framework/src/components/logger/logger.js]: Good from my end.

File: framework/src/controller/application.js (hunk 11 above)
Comment (nazarhussain) [reply framework/src/controller/application.js]: Let's create a seperate issue to refactor the coding style when we have a documented one. And then change all the code.

**Comment by nazarhussain** [comment]
@diego-G The issue is in my list will be created shortly to refactor and verify cleanup process.

**Comment by MaciejBaj** [comment]
I have created the issue of missing documentation - https://github.com/LiskHQ/lisk/issues/2845.

**Pull request merged**